# Introduction Numpy Library
## By elham heidari

In [3]:
#Import math library
import math 

In [ ]:
#mehtod   shift +tab()

In [4]:
math.sqrt(25)

5.0

In [6]:
math.log(100)

4.605170185988092

In [7]:
math.sin(30)

-0.9880316240928618

Basic Stat Functions

In [8]:
#Import statistics module
import statistics as stat

In [9]:
s = [1,3,5,6,1.0,11.2]
s

[1, 3, 5, 6, 1.0, 11.2]

In [11]:
stat.mean(s)

4.533333333333333

In [13]:
stat.median(s)

4.0

In [15]:
stat.mode(s)

1

In [16]:
stat.stdev(s)

3.8505410875183066

In [17]:
stat.variance(s)

14.826666666666664

# Numpy
NumPy (Numerical Python) is one of the core packages for numerical computing in Python.

Pandas, Matplotlib, Statmodels and many other Scientific libraries rely on NumPy.

Check NumPy documentation: https://numpy.org/doc/stable/user/absolute_beginners.html

In [ ]:
#% pip install numpy

In [18]:
import numpy as np

In [22]:
#One-dimension array
p1 = np.array([3, 2.1, 5, 9.7, 4, 5.5,1,15,12])
p1

array([ 3. ,  2.1,  5. ,  9.7,  4. ,  5.5,  1. , 15. , 12. ])

In [20]:
type(p1)

numpy.ndarray

In [23]:
#Shape is attribute
p1.shape

(9,)

In [24]:
#Indexing
p1[0]

3.0

In [25]:
p1[1 : 4]

array([2.1, 5. , 9.7])

In [26]:
p1[[0, 3, 4]]

array([3. , 9.7, 4. ])

In [28]:
#Matrix
p2 = np.array([[11,1.0, 1], [2, 0, 4], [1, 4, 3]])
p2

array([[11.,  1.,  1.],
       [ 2.,  0.,  4.],
       [ 1.,  4.,  3.]])

In [29]:
#Shape
p2.shape

(3, 3)

In [28]:
#Indexing
p2[1, 2]

4.0

In [31]:
p2[2][1]

4.0

In [32]:
p2[1, :]

array([2., 0., 4.])

In [34]:
p2

array([[11.,  1.,  1.],
       [ 2.,  0.,  4.],
       [ 1.,  4.,  3.]])

In [33]:
p2[1 :, 1:]

array([[0., 4.],
       [4., 3.]])

In [35]:
#Data type
p2.dtype

dtype('float64')

In [36]:
#Modify an array
p1

array([ 3. ,  2.1,  5. ,  9.7,  4. ,  5.5,  1. , 15. , 12. ])

In [37]:
p1[1] = 0
p1

array([ 3. ,  0. ,  5. ,  9.7,  4. ,  5.5,  1. , 15. , 12. ])

In [39]:
p3 = np.zeros((2, 3), dtype = int)
p3

array([[0, 0, 0],
       [0, 0, 0]])

In [40]:
p4 = np.ones((3, 3), dtype = int)
p4

array([[1, 1, 1],
       [1, 1, 1],
       [1, 1, 1]])

In [41]:
#Create ranges with arange
p5 = np.arange(1, 20, 2)
p5

array([ 1,  3,  5,  7,  9, 11, 13, 15, 17, 19])

In [42]:
p6 = np.arange(5)
p6

array([0, 1, 2, 3, 4])

In [48]:
#Create floating-point ranges with linspace
#The NumPy linspace function creates sequences of evenly spaced values within a defined interval
p7 = np.linspace(1, 5, 30)
p7

array([1.        , 1.13793103, 1.27586207, 1.4137931 , 1.55172414,
       1.68965517, 1.82758621, 1.96551724, 2.10344828, 2.24137931,
       2.37931034, 2.51724138, 2.65517241, 2.79310345, 2.93103448,
       3.06896552, 3.20689655, 3.34482759, 3.48275862, 3.62068966,
       3.75862069, 3.89655172, 4.03448276, 4.17241379, 4.31034483,
       4.44827586, 4.5862069 , 4.72413793, 4.86206897, 5.        ])

In [47]:
p7.shape

(20,)

In [50]:
p7 = p7.reshape(5, 6)
p7

array([[1.        , 1.13793103, 1.27586207, 1.4137931 , 1.55172414,
        1.68965517],
       [1.82758621, 1.96551724, 2.10344828, 2.24137931, 2.37931034,
        2.51724138],
       [2.65517241, 2.79310345, 2.93103448, 3.06896552, 3.20689655,
        3.34482759],
       [3.48275862, 3.62068966, 3.75862069, 3.89655172, 4.03448276,
        4.17241379],
       [4.31034483, 4.44827586, 4.5862069 , 4.72413793, 4.86206897,
        5.        ]])

In [51]:
#Create a 3x3 array of normally distributed random values
# with mean 0 and standard deviation 1
np.random.seed(123)
p8 = np.random.normal(0, 1, (3, 3))
p8

array([[-1.0856306 ,  0.99734545,  0.2829785 ],
       [-1.50629471, -0.57860025,  1.65143654],
       [-2.42667924, -0.42891263,  1.26593626]])

In [52]:
#Create a 3x3 array of random integers in the interval [0, 10)
np.random.seed(123)
p9 = np.random.randint(0, 10, (3, 3))
p9

array([[2, 2, 6],
       [1, 3, 9],
       [6, 1, 0]])

In [53]:
np.max(p9)

9

In [54]:
np.mean(p9)

3.3333333333333335

In [55]:
np.std(p9)

2.8284271247461903

In [56]:
np.min(p9)

0

In [49]:
#Sum along rows (sum over columns)
np.sum(p9, axis = 0)

array([ 9,  6, 15])

In [50]:
#Sum along columns (sum over rows)
np.sum(p9, axis = 1)

array([10, 13,  7])

In [51]:
#Array operations
p9 + 5

array([[ 7,  7, 11],
       [ 6,  8, 14],
       [11,  6,  5]])

In [52]:
p9 * 2

array([[ 4,  4, 12],
       [ 2,  6, 18],
       [12,  2,  0]])

In [54]:
p2 + p9

array([[13.,  3.,  7.],
       [ 3.,  3., 13.],
       [ 7.,  5.,  3.]])

In [55]:
p2 * p9

array([[22.,  2.,  6.],
       [ 2.,  0., 36.],
       [ 6.,  4.,  0.]])

In [56]:
#Matrix multiplication
np.dot(p2, p9)

array([[29., 26., 75.],
       [28.,  8., 12.],
       [24., 17., 42.]])

In [57]:
#Transpose of a matrix
p9.T

array([[2, 1, 6],
       [2, 3, 1],
       [6, 9, 0]])

In [58]:
#Boolian operations
p9 > 2

array([[False, False,  True],
       [False,  True,  True],
       [ True, False, False]])

In [59]:
#  &: and
#  |: or
#  ~: not

p9[(p9 % 3 == 0) & (p9 > 4)]

array([6, 9, 6])

In [60]:
#axis = 0 : concatenate vertically
#axis = 1 : concatenate horizontally
np.concatenate((p2, p9), axis = 0)

array([[11.,  1.,  1.],
       [ 2.,  0.,  4.],
       [ 1.,  4.,  3.],
       [ 2.,  2.,  6.],
       [ 1.,  3.,  9.],
       [ 6.,  1.,  0.]])

# Assignment

Q3: Create a 4×5 numpy array of all True’s.

Q4: Create an array of integers from 10 to 50 with with step of 3. Save the array in a. Replace all even numbers in a with -1.

Q5: Create an one-dimensional array which includes odd numbers from 1 to 20.

a: Change all numbers divisible by 3 into 0.

b: Change the array into a 2x5 matrix. 

c: Convert all numbers of the third column into zero.
Q6:

a: Create an one-dimensional array, b1, which includes 9 numbers evenly spaced over 1.5 to 3.  

b: Create a 3x3 matrix, b2, which includes random numbers from Gaussian distribution with mean = 1, and sd = 0.75, seed = 1234.

c: Calculate the determinant of b2.

d: What is the transpose of matrix b2?

e: Convert b1 into a 3x3 matrix.

f: Calculate summation of b1 and b2.

g: Calculate matrix multiplication of b1 and b2.

h: Call all numbers of b2 which are greater than mean of b2.

i: Scale matrix b2 using min-max scaling method.
End of Code